### PR TITLE
[GOBBLIN-1784] Only clean dags from the dag manager if a flow event is received

### PR DESCRIPTION
**Pull request by Will-Lo**
Dear Gobblin maintainers,

Please accept this PR. I understand that it will not be reviewed until I have checked off all the steps below!

### JIRA
- [x] My PR addresses the following [Gobblin JIRA](https://issues.apache.org/jira/browse/GOBBLIN/) issues and references them in the PR title. For example, "[GOBBLIN-XXX] My Gobblin PR"
    - https://issues.apache.org/jira/browse/GOBBLIN-1784

### Description
- [x] Here are some details about my PR, including screenshots (if applicable):
Currently, the DagManager will poll a dag until it receives that the *job* has been completed/failed/cancelled, and then emit the flow event and clean up the dag. However, on service restart or crash, it is possible that the flow event never gets emitted. Since the dag is already cleaned up, this job status becomes stuck, and will never progress.

To address this scenario, this PR only cleans up dags when the cleanup function polls that the *flow status* is completed. If it does not receive the flow status that the job is completed within 3 minutes of the dag finishing all of its jobs, then it will re-emit the flow event in case it was lost. 

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
  - Modified existing tests to the DagManagerTest class to ensure that the cleanup is working as expected on flow events

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### CODE REVIEW

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 1]
```diff
@@ -1117,67 +1107,90 @@ private boolean hasRunningJobs(String dagId) {
...
+          this.dagIdstoClean.add(dagId);
+          if (dag.getFlowEvent() == null) {
+            // If the dag flow event is not set, then it is successful
+            dag.setFlowEvent(TimingEvent.FlowTimings.FLOW_SUCCEEDED);
           } else {
-            dagManagerMetrics.emitFlowSuccessMetrics(DagManagerUtils.getFlowId(this.dags.get(dagId)));
+            addFailedDag(dagId, dag);
           }
-          log.info("Dag {} has finished with status {}; Cleaning up dag from the state store.", dagId, status);
+          log.info("Dag {} has finished with status {}; Cleaning up dag from the state store.", dagId, dag.getFlowEvent());
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1139]: not cleaning up dag at this point, so maybe change the log info here?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 2]
```diff
@@ -1117,67 +1107,90 @@ private boolean hasRunningJobs(String dagId) {
...
+              break;
+            case TimingEvent.FlowTimings.FLOW_CANCELLED:
+              this.dagManagerMetrics.emitFlowSlaExceededMetrics(flowId);
+              this.dagManagerMetrics.conditionallyMarkFlowAsState(flowId, FlowState.FAILED);
+              break;
+            default:
+              log.warn("Unexpected flow event {} for dag {}", dag.getFlowEvent(), dagId);
+          }
+          cleanUpDag(dagId);
+          cleanedDags.add(dagId);
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1170]: why do we remove it from dagIdstoClean in a separate loop instead of directly doing it here?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 3]
```diff
@@ -1117,67 +1107,90 @@ private boolean hasRunningJobs(String dagId) {
...
+            case TimingEvent.FlowTimings.FLOW_SUCCEEDED:
+              this.dagManagerMetrics.emitFlowSuccessMetrics(flowId);
+              this.dagManagerMetrics.conditionallyMarkFlowAsState(flowId, FlowState.SUCCESSFUL);
+              break;
+            case TimingEvent.FlowTimings.FLOW_FAILED:
+              this.dagManagerMetrics.emitFlowFailedMetrics(flowId);
+              this.dagManagerMetrics.conditionallyMarkFlowAsState(flowId, FlowState.FAILED);
+              break;
+            case TimingEvent.FlowTimings.FLOW_CANCELLED:
+              this.dagManagerMetrics.emitFlowSlaExceededMetrics(flowId);
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1163]: We seem to have an assumption that canceled flow must be because of sla exceed, is that always true?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java [hunk 4]
```diff
@@ -1048,9 +1049,6 @@ private void submitJob(DagNode<JobExecutionPlan> dagNode) {
     /**
      * Method that defines the actions to be performed when a job finishes either successfully or with failure.
      * This method updates the state of the dag and performs clean up actions as necessary.
-     * TODO : Dag should have a status field, like JobExecutionPlan has. This method should update that field,
```
Comment (ZihanLi58) [inline gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1051]: Seems we remove failedDagIdsFinishRunning, failedDagIdsFinishAllPossible in this PR as well. But I don't have enough context for what these two are used for right now and how current implementation can support the same use case. 
Should we separate this into a separate PR? Or at lease, please describe the change in PR description.

**Review by ZihanLi58 — CHANGES_REQUESTED** [review]

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java (hunk 2 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1170]: There's a concurrent modification exception due to the way how the dag Ids are stored, could store them in a map instead?

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java (hunk 4 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1051]: So those behaviors still stay, but i do the processing during cleanup since each dag stores a reference to their status so those sets are no longer needed.

File: gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java (hunk 1 above)
Comment (Will-Lo) [reply gobblin-service/src/main/java/org/apache/gobblin/service/modules/orchestration/DagManager.java:1139]: I'll move it lower to when the Dag is actually cleaned up, since this log is used for some analysis on number of flows with certain statuses

**Review by ZihanLi58 — APPROVED** [review]
+1
